### PR TITLE
Run Examples in CI. 

### LINKED ISSUE #571: "Parametric sweep" in demo/demo.ipynb fails (v0.4.1)
**Describe the bug**
Using MontePy 0.4.1 when running the example script (`demo/demo.ipynb`) fails in the final cell when attempting to open a file that does not exist: `parametric/pin_cell_0.35r_1.0p.imcnp`. The sub-directory `parametric` does not exist, if that sub-directory is manually generated by the user then things proceed without a hitch. 

My guess is this is a Mac only bug 

**To Reproduce**

Run the final cell in in `/demo/demo.ipynb`

**Error Message (if any)**

```
---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
Cell In[10], line 17
     15 for plane in box_walls:
     16     plane.location = (pitch / 2) * (-1 if plane.location < 0 else 1)
---> 17 problem.write_to_file(f\"parametric/pin_cell_{pin_radius}r_{pitch}p.imcnp\")

File /opt/homebrew/Caskroom/mambaforge/base/envs/mcdc-env/lib/python3.11/site-packages/montepy/mcnp_problem.py:529, in MCNP_Problem.write_to_file(self, file_path, overwrite)
    514 def write_to_file(self, file_path, overwrite=False):
    515     \"\"\"
    516     Writes the problem to a file.
    517 
   (...)
    527     :raises IsADirectoryError: if the path given is actually a directory.
    528     \"\"\"
--> 529     return self.write_problem(file_path, overwrite)

File /opt/homebrew/Caskroom/mambaforge/base/envs/mcdc-env/lib/python3.11/site-packages/montepy/mcnp_problem.py:507, in MCNP_Problem.write_problem(self, destination, overwrite)
    505 elif isinstance(destination, (str, os.PathLike)):
    506     new_file = MCNP_InputFile(destination, overwrite=overwrite)
--> 507     with new_file.open(\"w\") as fh:
    508         self._write_to_stream(fh)
    509 else:

File /opt/homebrew/Caskroom/mambaforge/base/envs/mcdc-env/lib/python3.11/site-packages/montepy/input_parser/input_file.py:133, in MCNP_InputFile.open(self, mode, encoding, replace)
    129     if os.path.isdir(self.path):
    130         raise IsADirectoryError(
    131             f\"{self.path} is a directory, and cannot be overwritten.\"
    132         )
--> 133 self._fh = open(self.path, mode, encoding=encoding)
    134 return self

FileNotFoundError: [Errno 2] No such file or directory: 'parametric/pin_cell_0.35r_1.0p.imcnp'"
```

**Version**

 - Version 0.4.1
 - Python 3.12.6

**Additional context**

This is part of the [PyOpenSci review](https://github.com/pyOpenSci/software-submission/issues/205)!

### LINKED ISSUE #575: Run Examples in CI
**Is your feature request related to a problem? Please describe.**
Example code is not tested automatically. It can fall out of date without anyone noticing, or it might not run out of the box.

**Describe the solution you'd like**
During CI, add an `example-test` that at least makes sure that the examples and demos execute.

**Additional context**
In #571, a demo does run without manual intervention.

In #559, @MicahGale suggests adding `doctest`, which is relevant.

### PR DESCRIPTION
**Pull request by MicahGale**
# Description

This implemented testing of (almost) all example code blocks. This includes:

* `README.md`
* `doc/source/*.rst`
* All docstring examples (there are none currently)
* `demo/*.ipynb`

This also fixed all bugs found. The only code that wasn't tested is ones where the output are multiline statements. This is too complicated to work with `sphinx-doctest`. 

Fixes #575, fixes #571

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

**Comment by MicahGale** [comment]
Using [doctest](https://docs.python.org/3/library/doctest.html), [sphinx-doctest](https://www.sphinx-doc.org/en/master/usage/extensions/doctest.html), and [papermill](https://pypi.org/project/papermill/) to run these tests.

**Comment by MicahGale** [comment]
` phmutest README.md --replmode --log` https://stackoverflow.com/a/76924724

**Comment by MicahGale** [comment]
I'm thinking we should override the changelog test.

File: doc/source/starting.rst [hunk 1]
```diff
@@ -519,25 +574,40 @@ As with OpenMC, the set logic operations have been mapped to python's bit logic
...
    cladding = ~gas_gap & ~slug_half_space & +bottom_plane & -top_plane & -clad_od
-
    # make weird multi-part cell
-  slugs = (+bottom_plane & -top_plane & -fuel_cylinder) | (+bottom_plane & -top_plane & -other_fuel)
+   slugs = (+bottom_plane & -top_plane & -fuel_cylinder) |  (+bottom_plane & -top_plane & -other_fuel)
 
 .. note::
+
   MontePy does not check if the geometry definition is "rational".
   It doesn't check for being finite, existant (having any volumen at all), or being infinite.
```
Comment (tjlaboss) [inline doc/source/starting.rst:612]: > volumen

File: doc/source/starting.rst [hunk 2]
```diff
@@ -519,25 +574,40 @@ As with OpenMC, the set logic operations have been mapped to python's bit logic
...
 .. note::
+
    When you combine two half-spaces with a logical operator you create a new half-space.
    In this case the concept of a side becomes much more about "in" and "out".
 
 .. note::
+
    Half-spaces need not be contiguous.
 
 Order of precedence and grouping is automatically handled by python so you can easily write complicated geometry in one-line.
```
Comment (tjlaboss) [inline doc/source/starting.rst:585]: capitalize Python

File: doc/source/starting.rst [hunk 3]
```diff
@@ -519,25 +574,40 @@ As with OpenMC, the set logic operations have been mapped to python's bit logic
...
-
+.. testcode::
+
+   # build blank surfaces 
+   bottom_plane = montepy.surfaces.axis_plane.AxisPlane()
+   top_plane = montepy.surfaces.axis_plane.AxisPlane()
+   fuel_cylinder = montepy.surfaces.cylinder_on_axis.CylinderOnAxis()
+   clad_cylinder = montepy.surfaces.cylinder_on_axis.CylinderOnAxis()
+   clad_od = montepy.surfaces.cylinder_on_axis.CylinderOnAxis()
+   other_fuel = montepy.surfaces.cylinder_on_axis.CylinderOnAxis()
```
Comment (tjlaboss) [inline doc/source/starting.rst:595]: So, I get that these objects are instantiated without surface coefficients, but it makes for a weird example not using any.

File: montepy/cell.py [hunk 4]
```diff
@@ -637,9 +637,8 @@ def __repr__(self):
                 ret += "atom/b-cm\n"
             else:
                 ret += "g/cc\n"
-        for surface in self._surfaces:
-            ret += str(surface) + "\n"
-        ret += "\n"
+        surf_strs = [str(s) for s in self.surfaces]
+        ret += "\n".join(surf_strs)
```
Comment (tjlaboss) [inline montepy/cell.py:641]:
```python
ret += "\n".join([str(s) for s in self.surfaces])
```

File: doc/source/starting.rst [hunk 5]
```diff
@@ -1,6 +1,10 @@
 Getting Started with MontePy
 ============================
 
+.. testsetup:: *
+
+   import montepy
+
 MontePy is a python API for reading, editing, and writing MCNP input files.
```
Comment (tjlaboss) [inline doc/source/starting.rst:8]: capitalize Python

File: doc/source/starting.rst [hunk 6]
```diff
@@ -1,6 +1,10 @@
...
 It does not run MCNP nor does it parse MCNP output files.
 The library provides a semantic interface for working with input files, or our preferred terminology: problems.
```
Comment (tjlaboss) [inline doc/source/starting.rst:10]:
```
MontePy is a Python API for reading, editing, and writing MCNP input files.
The library provides a semantic interface for working with input files ("MCNP problems").
It does not run MCNP, nor does it parse MCNP output files.
```

**Review by tjlaboss — CHANGES_REQUESTED** [review]
Changelog test ok. 

This is a testing and documentation update. We should make the following changes:

File: doc/source/starting.rst (hunk 3 above)
Comment (MicahGale) [reply doc/source/starting.rst:595]: Should I add coefficients?

Comment (tjlaboss) [reply doc/source/starting.rst:595]: yes.

**Review by tjlaboss — APPROVED** [review]

**Pull request merged**